### PR TITLE
Default interface implementations of static virtual methods - ECMA-335 augments

### PR DESCRIPTION
**Pull request by trylek**
This change proposes a number of additions and changes to the
ECMA-335 augments with regard to default interface implementations
of static virtual methods. For now I have kept the two sections
separate (Default Interface Methods vs. Virtual Static Methods)
to avoid excessive churn even though there's obvious overlap;
after all, both sections basically amount to "suggestions for a
future editor of the actual ECMA-335 spec" so that keeping them
separate is not a big deal.

(*) At this point I didn't attempt to construct a C# example
demonstrating default interface SVM implementations;

(*) Both sections contain suggestions for modifying the method
resolution algorithm described in II.12.2; each version has its
own merits - the Default Interface version is more focused on
describing the gist of the algorithm change whereas the SVM
version is attempting to propose a complete formulation of the
updated algorithm. I'm not sure what is the ideal way to approach
those.

Thanks

Tomas

### CODE REVIEW

**Comment by AlekseyTs** [comment]
I'll let @davidwrighton to review first.

File: docs/design/specs/Ecma-335-Augments.md
Comment (davidwrighton) [inline docs/design/specs/Ecma-335-Augments.md:447]: There should also be a clause about how "call" is extended to allow throwing `AmbiguousImplementationException`/`EntryPointNotFoundException`. It should be just like the rest of these.

**Review by davidwrighton — CHANGES_REQUESTED** [review]
Please fix the missing scenario around handling of the `call` opcode.

File: docs/design/specs/Ecma-335-Augments.md
Comment (trylek) [reply docs/design/specs/Ecma-335-Augments.md]: Thanks for pointing that out, added in 2nd commit.

**Review by davidwrighton — APPROVED** [review]

**Comment by davidwrighton** [comment]
I'd like to see an approving review from Aleksey before committing.

File: docs/design/specs/Ecma-335-Augments.md
Comment (AlekseyTs) [inline docs/design/specs/Ecma-335-Augments.md:402]: It looks like this is formatted as a sub-bullet under the "Interfaces are now allowed to have instance methods" bullet. Perhaps it should be on the same level instead.

File: docs/design/specs/Ecma-335-Augments.md
Comment (AlekseyTs) [inline docs/design/specs/Ecma-335-Augments.md:437]: > implicitly implemented by the type  [](http://example.com/codeflow?start=91&length=36)

I am not sure how to interpret the "implicitly implemented interface" term. Are there more than one way for a type to implement an interface? I see that this terminology was previously used for instance methods. So, this is more like a curiosity question.

File: docs/design/specs/Ecma-335-Augments.md
Comment (AlekseyTs) [inline docs/design/specs/Ecma-335-Augments.md:550]: >default implementation [](http://example.com/codeflow?start=4&length=22)

Is it appropriate to call implementations in a derived interface a "default implementation"? I usually use term "default implementation" for a method that is implemented at the point of its declaration.

Comment (AlekseyTs) [reply docs/design/specs/Ecma-335-Augments.md:550]: I.e. declared virtual vs. abstract

Comment (AlekseyTs) [inline docs/design/specs/Ecma-335-Augments.md:550]: >abstract static virtual method [](http://example.com/codeflow?start=35&length=30)

This doesn't apply to non-abstract virtual methods?

**Comment by AlekseyTs** [comment]
Done with review pass (commit 9)

File: docs/design/specs/Ecma-335-Augments.md
Comment (davidwrighton) [reply docs/design/specs/Ecma-335-Augments.md:437]: Yes. An interface can be implemented explicitly (where the interface exists in the interfaceimpl table for the type directly) or implicitly (where the interface is implemented, but only because the base type implemented the interface). This doesn't really matter for the semantics when Roslyn generates code, as Roslyn always redeclares all of the interfaces, but at runtime we have to handle the case where the base type has been changed since the derived type was had its IL generated by Roslyn. For... reasons, the behavior is slightly different.

File: docs/design/specs/Ecma-335-Augments.md
Comment (trylek) [reply docs/design/specs/Ecma-335-Augments.md:402]: Thanks Aleksey for spotting, I originally did think it was intentional but I think you're right, especially considering the bullet list only has four items, the two-level structure doesn't make any sense.

File: docs/design/specs/Ecma-335-Augments.md
Comment (trylek) [reply docs/design/specs/Ecma-335-Augments.md:550]: That is an interesting question. In another part of the spec augment (the bulleted list you commented on), I kept the pre-existing formulation *Interface methods are allowed to MethodImpl other interface methods the interface _requires_ (but we require the `MethodImpl`s to be final to keep things simple) - i.e. an interface is allowed to provide (or override) an implementation of another interface's method
  (instance or static virtual)*. I presume that the reference to `MethodImpl`s being *final* is supposed to mean that the SVM cannot be overridden multiple times at different places in the type hierarchy; this would likely also mean that a non-abstract static virtual method (i.e. one with implementation supplied at the point of declaration) cannot be overridden anywhere basically defeating the purpose of the `virtual` keyword. I don't see the need for this requirement and especially with default interface implementations of SVMs it no longer make much sense as we must traverse the interface hierarchies in well-defined order to achieve consistent results. I have updated the spec to this more general model; we'll see what @davidwrighton has to say on the subject.

Comment (trylek) [reply docs/design/specs/Ecma-335-Augments.md]: I'm not 100% sure about the precise terminology here. I was under the impression that the term "default interface implementations" was supposed to emphasize the deviation from the traditional notion that interfaces are just abstract API contracts. In general I like your point so for now I deleted the word `default` from the formulation.

**Comment by trylek** [comment]
I believe I have addressed all Aleksey's PR feedback; in some of these cases I would welcome additional feedback from @davidwrighton. Can you please take another look when you have a chance?

**Comment by AlekseyTs** [comment]
@trylek Force pushes and rewriting commits in PRs under review have a negative impact on the review process because they prevent reviewing changes in an incremental fashion.

**Comment by trylek** [comment]
Apologies for the inconvenience, I'll try to be more considerate w.r.t. future pushes on my PRs.

**Comment by danmoseley** [comment]
@trylek are you still working on this PR? (going through old PR's in my area - meta)

If it's on hold, let's close for now.

**Comment by trylek** [comment]
I'm closing this for now, it's far lower priority than other release-related tasks.

**Pull request closed without merging**
